src/libteam/0005-libteam-Add-warm_reboot-mode.patch

```diff
@@ -0,0 +1,354 @@
+-		if (!ifinfo->master_ifindex)
++		if (!ifinfo->orig_addr_updated) {
+ 			memcpy(ifinfo->orig_hwaddr, hwaddr, hwaddr_len);
++                        ifinfo->orig_addr_updated = true;
++                }
+ 		set_changed(ifinfo, CHANGED_HWADDR);
+ 	}
+ }
+diff --git a/teamd/teamd.c b/teamd/teamd.c
+index c987333..37c2fb3 100644
+--- a/teamd/teamd.c
++++ b/teamd/teamd.c
+@@ -116,7 +116,9 @@ static void print_help(const struct teamd_context *ctx) {
+             "    -D --dbus-enable         Enable D-Bus interface\n"
+             "    -Z --zmq-enable=ADDRESS  Enable ZeroMQ interface\n"
+             "    -U --usock-enable        Enable UNIX domain socket interface\n"
+-            "    -u --usock-disable       Disable UNIX domain socket interface\n",
++            "    -u --usock-disable       Disable UNIX domain socket interface\n"
++            "    -w --warm-reboot         Warm-reboot startup mode\n"
++            "    -L --lacp-directory      Directory for saving lacp pdu dumps\n",
+             ctx->argv0);
+ 	printf("Available runners: ");
+ 	for (i = 0; i < TEAMD_RUNNER_LIST_SIZE; i++) {
+@@ -149,10 +151,12 @@ static int parse_command_line(struct teamd_context *ctx,
+ 		{ "zmq-enable",		required_argument,	NULL, 'Z' },
+ 		{ "usock-enable",	no_argument,		NULL, 'U' },
+ 		{ "usock-disable",	no_argument,		NULL, 'u' },
++		{ "warm-reboot",        no_argument,		NULL, 'w' },
++		{ "lacp-directory",     required_argument,	NULL, 'L' },
+ 		{ NULL, 0, NULL, 0 }
+ 	};
+ 
+-	while ((opt = getopt_long(argc, argv, "hdkevf:c:p:groNt:nDZ:Uu",
++	while ((opt = getopt_long(argc, argv, "hdkevf:c:p:groNt:nDZ:UuwL:",
+ 				  long_options, NULL)) >= 0) {
+ 
+ 		switch(opt) {
+@@ -230,6 +234,12 @@ static int parse_command_line(struct teamd_context *ctx,
+ 		case 'u':
+ 			ctx->usock.enabled = false;
+ 			break;
++		case 'w':
++			ctx->warm_reboot = true;
++			break;
++		case 'L':
++			ctx->lacp_directory = strdup(optarg);
++			break;
+ 		default:
+ 			return -1;
+ 		}
+@@ -384,8 +394,14 @@ static int teamd_run_loop_run(struct teamd_context *ctx)
+ 			if (err != -1) {
+ 				switch(ctrl_byte) {
+ 				case 'q':
++				case 'w':
+ 					if (quit_in_progress)
+ 						return -EBUSY;
++					if (ctrl_byte == 'w') {
++						ctx->keep_ports = true;
++						ctx->no_quit_destroy = true;
++						teamd_ports_flush_data(ctx);
++					}
+ 					teamd_refresh_ports(ctx);
+ 					err = teamd_flush_ports(ctx);
+ 					if (err)
+@@ -428,6 +444,12 @@ void teamd_run_loop_quit(struct teamd_context *ctx, int err)
+ 	teamd_run_loop_sent_ctrl_byte(ctx, 'q');
+ }
+ 
++static void teamd_run_loop_quit_w_boot(struct teamd_context *ctx, int err)
++{
++	ctx->run_loop.err = err;
++	teamd_run_loop_sent_ctrl_byte(ctx, 'w');
++}
++
+ void teamd_run_loop_restart(struct teamd_context *ctx)
+ {
+ 	teamd_run_loop_sent_ctrl_byte(ctx, 'r');
+@@ -694,6 +716,10 @@ static int callback_daemon_signal(struct teamd_context *ctx, int events,
+ 		teamd_log_warn("Got SIGINT, SIGQUIT or SIGTERM.");
```

Comment:
>SIGQUIT [](start = 31, length = 7)

We may reuse sigterm instead of sigusr1.
If lacp-directory is provided, we can save state, otherwise just normal quit.

Reply:
The warm stop is not about file saving only. It includes other activities, like remove portchannel interface, put down physical ports